action.yml

```diff
@@ -77,10 +77,149 @@ runs:
+              throw "LocalPath parameter is specified, but the file does not exist"
+            }
+            $FilePath = $LocalPath
+          } else {
+            $FileName = [System.IO.Path]::GetFileName($URL)
+            $FilePath = Invoke-Download -URL $URL -Path "${env:Temp}\$FileName"
+          }
+
+          if ($ExpectedSHA256) {
+            $Hash = (Get-FileHash -Path $FilePath -Algorithm SH256).Hash
+            Verify-Checksum -Actual $Hash -Expected $ExpectedSHA256
+          }
+
+          if (-not $Type) {
+            $Type = ([System.IO.Path]::GetExtension($FilePath)).Replace(".", "").ToUpper()
+          }
+
+          switch ($Type) {
+            "EXE" {
+              $InstallArgs = $ExtraInstallArgs
+            }
+            "MSI" {
+              if ($ExtraInstallArgs) {
+                $InstallArgs = @("/i", $FilePath, "/qn", "/norestart") + $ExtraInstallArgs
+              } else if (-not $InstallArgs) {
+                Write-Host "Using default MSI arguments: /i, /qn, /norestart"
+                $InstallArgs = @("/i", $FilePath, "/qn", "/norestart")
+              }
+              $FilePath = "msiexec.exe"
+            }
+            default {
+              throw "Unknown installer type (${Type}), please specify via `-Type` parameter"
+            }
+          }
+
+          $StartTime = Get-Date
+          Write-Host "Starting Install..."
+          try {
+            $Process = Start-Process -FilePath $FilePath -ArgumentList $InstallArgs -Wait -PassThru
+            $ExitCode = $Process.ExitCode
+            $Duration = [math]::Round(($(Get-Date) - $StartTime).TotalSeconds, 2)
+            switch ($ExitCode) {
+              0 { Write-Host "Installation successful in $Duration seconds" }
+              3010 { Write-Host "Installation successful in $Duration seconds; reboot required"}
+              default {
+                Write-Host "Installation process returned unexpected exit code: $ExitCode"
+                Write-Host "Time elapsed: $Duration seconds"
+                exit $ExitCode
+              }
+            }
+          } catch {
+            $Duration = [math]::Round(($(Get-Date) - $StartTime).TotalSeconds, 2)
+            Write-Host "Installation failed after $Duration seconds"
+          }
+        }
+
         if ("${{ steps.validation.outputs.use_custom_url }}" -eq "1") {
-          Install-Binary -FilePath "${{ inputs.release-asset-name }}" -ArgumentList ("-q")
+          Invoke-Installation -FilePath "${{ inputs.release-asset-name }}" -ArgumentList ("-q")
         } else {
-          Install-Binary -Url "https://download.swift.org/${{ inputs.branch }}/windows10/swift-${{ inputs.tag }}/swift-${{ inputs.tag }}-windows10.exe" -Name "installer.exe" -ArgumentList ("-q")
+          Invoke-Installation -URL "https://download.swift.org/${{ inputs.branch }}/windows10/swift-${{ inputs.tag }}/swift-${{ inputs.tag }}-windows10.exe" -Name "installer.exe" -ArgumentList ("-q")
```

Comment:
`-ArgumentList` isn't a param for `Invoke-Installation`?

Reply:
Ah, right, that is `InstallArgs` now.  I think I prefer `ArgumentList` so I will rename that.